### DISCUSSION TITLE
[DONE] New expressions: "Object.GetXByAngleAndDistance(angle, distance)" and "Object.GetYByAngleAndDistance(angle, distance)"

### DISCUSSION THREAD

**Discussion by tristanbob** [4ian/GDevelop#2367] [Archive] [answered]
## Description

A gamedev knows a vector (angle and distance) and they want to know the endpoint position of the vector.  This will be useful for any function that requires a position as input, such as position Tween and Pathfinding.  This will make it easy to use tweens for knockback or dashes in any direction.

## Solution suggested

Create two new expressions:

Object.VectorX(angle, distance) = Given an angle and distance, return the X position of the resulting vector.
Object.VectorY(angle, distance) = Given an angle and distance, return the Y position of the resulting vector.

This can be calculated using trigonometry functions:  

Object.VectorX(angle, distance) = return Object.X() + (distance * cos(ToRad(angle)))
Object.VectorY(angle, distance) = return Object.Y() + (distance * sin(ToRad(angle)))

**Comment by 4ian** [comment]
This would be surely interesting. Note that we have the reverse already:

- For objects:
  - get the angle: `Object.AngleToObject`, `Object.AngleToPosition`
  - get the distance: `Object.Distance` (should have been called `DistanceToObject`... no big deal), `Object.DistanceToPosition` (and the squared equivalent: `SqDistance`, `SqDistanceToPosition`).
- Expressions:
  - angle: `AngleBetweenPositions`
  - distance: `DistanceBetweenPositions`
  
Note that to be useful in most cases, the Angle and Distance for objects are from the **object center**.

So what you're suggesting is to offer the reverse operation.

I'm not sure about two things. 
1. The naming. Basically, we translate from "polar coordinates" to "cartesian coordinates". In GDevelop, we don't talk about cartesian coordinates but just about "X/Y" and we don't tell usually "Polar" coordinates, but just "Distance and angle to somewhere/something". So should we should call this:
  - `VectorX`/Y
  - `PolarVectorX`/Y
  - something else? `DistanceAndLengthX`/Y? 
2. The Object.VectorX/Y should be, to be consistent with what's existing, calculated from the **object center**.

**Reply by tristanbob** [reply]
> 1. The Object.VectorX/Y should be, to be consistent with what's existing, calculated from the **object center**.

Do you mean this should use PointX("Center")?  If so, it will restrict this function to only working with Sprites (which might be OK).  I'm not sure how else we can calculate the center, since people can change the origin point.

**Comment by Silver-Streak** [comment]
If we're looking to keep the same naming structure, I'd think "PositionByAngleAndDistance", "Object.PositionByDistanceAndAngle" are both accurate but also relatively long.

**Comment by 4ian** [comment]
Yes that would work even if a bit verbose `PositionXByDistanceAndAngle`/`PositionYByDistanceAndAngle `. 
Could also by without abbreviation: `XByDistanceAndAngle`/`YByDistanceAndAngle` but maybe too implicit now..

**Comment by Triarite** [comment]
I can personally say that adding a Vector expression or whatever you prefer to call it would prove extremely useful to my projects. I, of course, can't speak for others. But it doesn't seem like this would be such a difficult expression to add, and considering how helpful it could prove to be I'm not sure why it can't be added, and sooner rather than later at that.

**Comment by Silver-Streak** [comment]
@4ian Just chiming back in on this as it's come up again on the discord.

I think  X/YByDistanceAndAngle, or maybe GetX/YByDistanceAndAngle works slightly better and is shorter than Position, but less implicit than x/y?

**Reply by tristanbob** [reply]
I prefer listing the angle first since that is what seems most natural: "If I head in this direction for 100 units, where will I end up?"

```
GetXByAngleAndDistance()
GetYByAngleAndDistance()
```

**Reply by 4ian** [reply]
Usually expressions in GDevelop don't have Get prefix, because everything is a Get.

So would rather be:
  XByAngleAndDistance(angle, distance)
  YByAngleAndDistance(angle, distance)
  
The angle is indeed first in the action to set a force, so we'll probably keep this order here to.
I think I'm fine with this.

Last question, is it better to have `XByAngleAndDistance` or `XFromAngleAndDistance`? The "by" must be conveying better the idea of computing a new position from an angle and a distance (but in this sentence, I used "from").

**Reply by Silver-Streak** [reply]
Looking through http://wiki.compilgames.net/doku.php/gdevelop5/all-features/expressions-reference, there are currently 0 expressions with "By", and many expressions with "From", so From is more standard even if those expressions aren't position related, I suppose?

**Comment by Bouh** [comment]
I can add these new expressions `XFromAngleAndDistance` & `YFromAngleAndDistance` if you want, i just need to know: 

- The text to show in the expression editor (Like: Angle between two objects)
- The long description (Like: Compute the angle between two objects. If you need the angle to an arbitrary position, use AngleToPosition.)

I can take care of the other points 👍

**Reply by tristanbob** [reply]
Perhaps this?

Short:
X position from angle and distance

Long:
Compute the X position when given an angle and distance relative to the starting object.

**Reply by Silver-Streak** [reply]
For long, I would do
"Compute the X position when given an angle and distance relative to the starting object. This is also known as a 2D Vector"

Just because people may be looking for it based off the Vector name.

This actually makes me wonder if the expression should be 2DVectorX/Y? I feel like that may not make sense to some folks, though.

**Reply by 4ian** [reply]
> X position from angle and distance

Sounds good.

For the long description, I would even say: "Compute the X position when given an angle and distance relative to the starting object. This is also known as getting the cartesian coordinates of a 2D vector, using its polar coordinates."

So we have "vector" there for people that want the full explanation, and a reasonable name for the rest of users.

I would prefer to keep `X/YFromAngleAndDistance` because while we're talking about "vectors" here (i.e: a pair of coordinates), it's not obvious when you read `2DVectorX(180, 200)` that the arguments are an angle and a length. On the contrary, in a traditional programming language, if I write `2DVector(180, 200)`, I would assume I'm creating a "2DVector" object with a x coordinate of 180 pixels and y coordinate of 200 pixels 😅
(and such a programming language could offer a static function like `2DVector.FromPolar(180, 200)` or `2DVector.FromAngleAndDistance(180, 200)`, which we simplify in GDevelop to `XFromAngleAndDistance(180, 200)` as we don't deal with vector objects).

**Reply by Bouh** [reply]
Here the pull request with what the result is in preview
https://github.com/4ian/GDevelop/pull/2558

**Reply by 4ian** [reply]
I have a sudden doubt about the naming. These expressions return the X and Y values from a starting point (the object center, or for the version without an object, the position 0;0) to a position that is specified by its angle and distance.
So we could name it: `XToAngleAndDistance`. But this looks like a conversion, while X is the result.
So we could name it too: `AngleAndDistanceToX`. But it sounds a bit weird.

So I think we can still keep `XFromAngleAndDistance` (X is the result, that's clear, and we get it from an angle and a distance).

Still writing this so you can all tell me if it makes sense.

**Reply by Bouh** [reply]
Your doubt create a doubt to me, if I add a free expression with `XFromAngleAndDistance` and an object expression with the same `XFromAngleAndDistance` it's ok or we have a potential problem of collision ?

**Reply by 4ian** [reply]
Yes it's no problem, object expressions are scoped to objects.

**Comment by Bouh** [comment, marked as answer]
@tristanbob These expressions has been added in [this pull request](https://github.com/4ian/GDevelop/pull/2558). You can try it in lastest [nightly builds](https://github.com/4ian/GDevelop/blob/master/newIDE/docs/Nightly-Builds-and-continuous-deployment.md).

Expressions are:

```
//Object expressions
object.XFromAngleAndDistance(angle, distance)
object.YFromAngleAndDistance(angle, distance)

//Free expressions
XFromAngleAndDistance(angle, distance)
YFromAngleAndDistance(angle, distance)

```




This conversation can be marked solved :)
